src/hotspot/share/opto/addnode.cpp

```diff
@@ -332,6 +332,21 @@ Node *AddINode::Ideal(PhaseGVN *phase, bool can_reshape) {
     }
   }
 
+  // Convert (x >>> rshift) + (x << lshift) into RotateRight(x, rshift)
+  if (Matcher::match_rule_supported(Op_RotateRight) &&
+      ((op1 == Op_URShiftI && op2 == Op_LShiftI) || (op1 == Op_LShiftI && op2 == Op_URShiftI)) &&
+      in(1)->in(1) == in(2)->in(1)) {
+    Node* rshift = op1 == Op_URShiftI ? in(1)->in(2) : in(2)->in(2);
+    Node* lshift = op1 == Op_URShiftI ? in(2)->in(2) : in(1)->in(2);
+    const TypeInt* rshift_t = phase->type(rshift)->isa_int();
+    const TypeInt* lshift_t = phase->type(lshift)->isa_int();
+    if (lshift_t != NULL && lshift_t->is_con() &&
+        rshift_t != NULL && rshift_t->is_con() &&
+        ((lshift_t->get_con() & 0x1F) == (32 - (rshift_t->get_con() & 0x1F)))) {
+      return new RotateRightNode(in(1)->in(1), phase->intcon(rshift_t->get_con() & 0x1F), TypeInt::INT);
+    }
+  }
+
```

Comment:
Even though existing code in that method seems to assume node's inputs can't be NULL, it's a good practice to protect against unexpected NULLs as that can happen when sub-graphs die during IGVN. So in(1)->in(1), in(1)->in(2), in(2)->in(2) need to be tested for NULL.

That logic and the one for AddLNode are almost identical. So it would be good to have it in a shared method. I've been adding helper methods to make that possible but not all of that code is in yet. Could you file a bug to revisit this issue later and assign it to me?

Reply:
Hi Roland,

Thanks for your feedback.

> Even though existing code in that method seems to assume node's inputs can't be NULL, it's a good practice to protect against unexpected NULLs as that can happen when sub-graphs die during IGVN. So in(1)->in(1), in(1)->in(2), in(2)->in(2) need to be tested for NULL.

Agree

> That logic and the one for AddLNode are almost identical. So it would be good to have it in a shared method. I've been adding helper methods to make that possible but not all of that code is in yet.

As this match rule is trivial enough, how about withdrawing these shared code in this PR for integrating backend first if your helper methods coming soon?  

> Could you file a bug to revisit this issue later and assign it to me?

Okay, it's on my queue now:P


-- Eric

Reply:
> As this match rule is trivial enough, how about withdrawing these shared code in this PR for integrating backend first if your helper methods coming soon?

https://github.com/openjdk/jdk/pull/2045 is the one I'm referring to. No idea when it's going to get in so I would suggest to move forward with the current change and revisit it later.

Reply:
Got it. The patch has been updated.